### PR TITLE
Warp and Logging and Application Transformer Demo

### PR DESCRIPTION
**Pull request by CMCDragonkai**
@Zachaccino @nzhang-zh can you help flesh this out?

### CODE REVIEW

File: src/Demo.hs [hunk 1]
```diff
@@ -32,30 +45,50 @@ newtype DemoT m a = DemoT
...
 -- connect the application monad transformer to this
 
+defaultEnv :: DemoEnv
+defaultEnv = DemoEnv (read "127.0.0.1") (read "8769")
+
+runDemo :: DemoEnv -> Demo a -> IO a
+runDemo env demo = do
+  runStderrLoggingT $ runReaderT (Demo.run demo) env
+
+myDemo :: Demo ()
```
Comment (nzhang-zh) [inline src/Demo.hs]: Does this count as connecting the application monad to the server? @CMCDragonkai

**Comment by CMCDragonkai** [comment]
I've cleaned this up and now have 2 executables produced from a `nix-build ./release.nix --attr application`.

This demonstrates how to use Warp with monad-logger and safe-exceptions and the ReaderT application transformer pattern, and also the usage of `data-default-class`.

The way I get the env variables applied to the default parameters is a bit bulky, I think there's a better way to do it.

I want to mention that I realized that warp uses IO at the end instead of MonadIO. Which sort of means that once you go into the warp context, you can't really use the same logging context as the others. Instead we end up "unwrapping" the logging immediately in IO. And this sort of makes sense in that the handling of a request happens in IO anyway, and at that point the logging code would have to be executed and directed to a particular exit point. This is currently hardcoded as `runStderrLoggingT`. It may be possible to put this function as part of the `DemoEnv`, so that way the ReaderT itself dictates where this will be, and then we would "ask" for the function and inject it into our `warpApp`.

@ksenia-portu @nzhang-zh @Zachaccino please review this code.

I'm merging this.

**Comment by CMCDragonkai** [comment]
Oh and `cabal v2-repl` is pretty useful. Make sure to use `:m *Demo` to get into the module context instead of `:m +Demo` which just "imports" the module.

**Comment by CMCDragonkai** [comment]
References:

* https://www.reddit.com/r/haskellquestions/comments/33c6bu/what_is_monadreader_for_when_we_have_readert_and/
* https://thoughtbot.com/blog/refactoring-to-a-monad-transformer-stack
* https://github.com/Holmusk/three-layer

**Pull request merged**